### PR TITLE
Kernel/aarch64: Flush data cache when updating page tables

### PR DESCRIPTION
**Pull request by FireFox317**
This is currently a draft as it relies on `Aarch64::Asm::flush_data_cache` present in #18876.

Found by running the aarch64 kernel bare metal on a Raspberry Pi 3B.

With these two commits, we now get to the point where the Scheduler starts, and we hang indefinitely.

### CODE REVIEW

File: Kernel/Arch/aarch64/Processor.cpp [hunk 1]
```diff
@@ -109,8 +109,12 @@ void Processor::initialize(u32)
         asm volatile("wfi");
 }
 
-void Processor::flush_tlb_local(VirtualAddress, size_t)
+void Processor::flush_tlb_local(VirtualAddress vaddr, size_t page_count)
 {
+    // NOTE: We must flush the data cache here, such that the updated page tables actually
```
Comment (nico) [inline Kernel/Arch/aarch64/Processor.cpp:114]: Unsolicited writing advice: always `s/note: //` as it's always information-free (why else would you write a comment).

(Optional nit, obvsly)

File: Kernel/Arch/aarch64/Processor.cpp [hunk 2]
```diff
@@ -109,8 +109,12 @@ void Processor::initialize(u32)
...
+    //       end up in main memory, and thus actually get used by the MMU.
+    Aarch64::Asm::flush_data_cache(vaddr.get(), 4096 * page_count);
```
Comment (nico) [inline Kernel/Arch/aarch64/Processor.cpp:116]: `s/4096/PAGE_SIZE` (or is that no available here?)

**Review by nico — APPROVED** [review]
\o/

File: Kernel/Arch/aarch64/Processor.cpp [hunk 3]
```diff
@@ -109,8 +109,12 @@ void Processor::initialize(u32)
...
+
     // FIXME: Figure out how to flush a single page
```
Comment (ADKaster) [inline Kernel/Arch/aarch64/Processor.cpp:118]: The code you added fixes the FIXME here, yes? Or no.

File: Kernel/Arch/aarch64/Processor.cpp (hunk 1 above)
Comment (konradekk) [reply Kernel/Arch/aarch64/Processor.cpp:114]: (i go for `XXX` if something is really not obvious, but i vaguely remember this tag was some kind of surprise here?)

File: Kernel/Arch/aarch64/Processor.cpp [hunk 4]
```diff
@@ -109,8 +109,12 @@ void Processor::initialize(u32)
         asm volatile("wfi");
 }
 
-void Processor::flush_tlb_local(VirtualAddress, size_t)
+void Processor::flush_tlb_local(VirtualAddress vaddr, size_t page_count)
```
Comment (konradekk) [inline Kernel/Arch/aarch64/Processor.cpp:112]: (it seems you use a different signature consistently in almost all the places; not sure if it begs for a inline proxy…? also `FlatPtr` should probably be use all along there—i guess that those are out of scope and weʼll just increase the crust a little! 😅)

File: Kernel/Arch/aarch64/Processor.cpp (hunk 1 above)
Comment (kleinesfilmroellchen) [reply Kernel/Arch/aarch64/Processor.cpp:114]: > (i go for `XXX` if something is really not obvious, but i vaguely remember this tag was some kind of surprise here?)

We don't use `XXX`, as `NOTE` it's superfluous and mostly found in older code.

File: Kernel/Arch/aarch64/Processor.cpp (hunk 2 above)
Comment (kleinesfilmroellchen) [reply Kernel/Arch/aarch64/Processor.cpp:116]: Should be available in all Serenity-only code; on Lagom it sometimes expands to a `sysconf` call.

File: Kernel/Arch/aarch64/Processor.cpp (hunk 1 above)
Comment (FireFox317) [reply Kernel/Arch/aarch64/Processor.cpp:114]: Sure, will remove the `NOTE`, however in Andreas latest video he was adding such a `NOTE` comment too (later removed it tho), but that was in LibWeb, so 🤷

File: Kernel/Arch/aarch64/Processor.cpp (hunk 2 above)
Comment (FireFox317) [reply Kernel/Arch/aarch64/Processor.cpp:116]: Yep, good catch, forgot to use that.

File: Kernel/Arch/aarch64/Processor.cpp (hunk 3 above)
Comment (FireFox317) [reply Kernel/Arch/aarch64/Processor.cpp:118]: Nope, that FIXME is about the TLB flushing below, which still flushes the complete TLB caches. I know it's kinda confusing, but there is a difference between the data cache and the TLB cache. The TLB cache just caches the result of virtual->physical translations, such that successive translations can use the cached result.

File: Kernel/Arch/aarch64/Processor.cpp (hunk 4 above)
Comment (FireFox317) [reply Kernel/Arch/aarch64/Processor.cpp:112]: Yeah, really the callsites should be using VirtualAddresses instead of raw pointers instead, but that's once for the :yakstack:

File: Kernel/Memory/Region.cpp [hunk 5]
```diff
@@ -225,6 +225,8 @@ bool Region::map_individual_page_impl(size_t page_index, RefPtr<PhysicalPage> pa
     if (!pte)
         return false;
 
+    ScopeGuard flush_tlb_guard = [pte] { Processor::flush_tlb_local(VirtualAddress((FlatPtr)pte), 1); };
```
Comment (konradekk) [inline Kernel/Memory/Region.cpp:228]: (it seems you use default `1` in other cases; not a big deal of course—just noting 😝)

**Review by kleinesfilmroellchen — APPROVED** [review]

**Comment by ADKaster** [comment]
@FireFox317  is this one still a draft? or were you wanting to add some additional commits to it.

**Comment by FireFox317** [comment]
Yeah, this is still a draft as the change still causes a crash when unmapping memory, and I haven't had the time to find a solution yet.
